### DISCUSSION TITLE
company-auto-commit considerations

### DISCUSSION THREAD

**Discussion by yugaego** [company-mode/company-mode#1264] [Ideas]
After having written a description for `company-auto-commit` and `company-auto-commit-chars` for the manual I failed to see any advantages in providing these two user options other than for backward compatibility. 
It seems to me, that a user may configure custom binding(s) to call `'company-complete-selection` (or a custom function that would call `company-complete-selection`) and reproduce behavior provided by the former pair.
Am I overlooking something?

Also (kinda bikeshedding follows), the names `company-auto-commit*` seems to be out of place according to the rest of the terminology used throughout the `Company` and don't clearly reflect what these options enable.
IMO, the previously used term "complete" (vs "commit") was the correct one in respect to Company. The term "auto" here rather means a "triggered" action. So, might be some variation of `company-trigger-complete` (`company-complete-selection-triggered` or `-force`), `company-complete-selection-chars` (`-on-chars`, `-chars-triggers`) could be a better name.
But, anyway, it probably would be an unhealthy act to change the names again at this stage?

**Summing up** the described two consideration, might be `company-auto-commit` feature could be planned for removal from Company?

**Comment by yugaego** [comment]
Addressing the first part of my hesitations: probably a good use case specifically for `company-auto-commit` would be using it with "syntax characters", to work nicely as an abstraction for different languages (syntax tables).

**Comment by dgutov** [comment]
Hi!

You are probably right about the name being a little out of place. But I grew really tired with people mistaking those variable with what the `auto-complete` package did/does. And thinking that they must set `company-auto-complete` to `t` for the automatic completion ("idle completion" in Company parlance) to work.

`company-trigger-complete` or some variation does sound closer to the existing terms. Is it too late to change?.. Maybe, but maybe not. I'm open to that idea. We could notify all users in all related issues, at least.

But regarding

> After having written a description for company-auto-commit and company-auto-commit-chars for the manual I failed to see any advantages in providing these two user options other than for backward compatibility.
It seems to me, that a user may configure custom binding(s) to call 'company-complete-selection (or a custom function that would call company-complete-selection) and reproduce behavior provided by the former pair.
Am I overlooking something?

...I would say yes. First, we do have the `company-quick-access-modifier` stuff, right? Even though the user could set up those bindings in their init file as well.

But more importantly, the `company-auto-commit` stuff only comes into play when the last input did not result in a valid completion. See `company--continue`. So if, for example, `company-auto-commit-chars` contains the dash character (or space, or paren, etc; some character), but when you type it, the current backend recognizes it as a prefix constituent, and suggests some completions too, the "auto commit" logic isn't going to fire. Completion will simply continue.

That's not something a user can easily set up in the keymap.

(Sorry for the late reply)

**Comment by yugaego** [comment]
Hi Dmitry (and congrats on the EmacsConf participation),

> I grew really tired with people mistaking those variable with what the auto-complete package did/does. And thinking that they must set company-auto-complete to t for the automatic completion ("idle completion" in Company parlance) to work.

Then I wasn't mistaken adding the following footnote to the manual's draft (BTW I'm going to push the updates quite soon):
> "The options company-auto-commit and company-auto-commit-chars used to be called company-auto-complete and company-auto-complete-chars respectively, which was in more accordance with the terminology given in this manual. But the resulting combination of the words ‘auto-complete’ present in those names made it seem the role of these user options was to configure Company’s auto-start behavior. Hence, it was chosen to rename the options to, hopefully, less confusing names."

> company-trigger-complete or some variation does sound closer to the existing terms. Is it too late to change?.. Maybe, but maybe not. I'm open to that idea. 

Okay, then I'm keeping it open on my agenda. The naming might turn out to be self-evident after settling with the terminology the manual puts forward. 

> But more importantly, the company-auto-commit stuff only comes into play when the last input did not result in a valid completion.

I got it more clearly now. Thank you for the explanation.

> (Sorry for the late reply)

No problem 😉

**Reply by dgutov** [reply]
> Hi Dmitry (and congrats on the EmacsConf participation),

Thanks!

> Okay, then I'm keeping it open on my agenda. The naming might turn out to be self-evident after settling with the terminology the manual puts forward.

Sounds good. :+1:
